File: credit.c
#include <stdio.h>
#include <stdlib.h>

int main()
{
// Program that prompts the user for a credit card number and then
// reports whether it is a valid American Express, MasterCard or Visa card number.

    long long card_number;
    int dig_count = 0;

   // Ask for credit card number
    do
    {
    printf("Enter your credit card: ");
    card_number = GetLongLong();

    }
    while (card_number < 0);

    //Get number of digits in card number
    long long temp = card_number;
    while (temp >= 1)
    {
        temp = temp / 10;
        dig_count++;
    }

    //Check if the card is valid according to the length
    if (!(dig_count == 13 || dig_count == 15 || dig_count == 16))
    {
        printf("INVALID\n");
        exit(1);
    }

    //Put the card number into array
    int number[dig_count];
    temp = card_number;

    for (int i = 0, j = dig_count-1; i < dig_count; i++)
    {
        number[j--] = temp % 10;
        temp /= 10;
    }

    //Get the sum of numbers, that are placed at the even places
    int summa = 0;
    int x;
    for (int i = 1; i < dig_count; i += 2)
    {
        x = number[i]*2;
        summa += x / 10 + x % 10;
    }

    //Adding the rest of numbers to the sum above
    for (int i = 0; i < dig_count; i+=2)
    {
        summa += number[i];
    }

    //Check if the last digit is 0
    if(summa % 10 != 0)
    {
        printf("INVALID");
        exit(1);
    }

    //Check the initial digits and print the result
    x = number[0] * 10 + number[1];

    switch(dig_count)
    {
        case 13: if(number[0] == 4) printf("VISA"); break;
        case 15: if(x == 34 || x == 37) printf("AMERICAN"); break;
        case 16: if(number[0] == 4) printf("VISA");
                 else if(x > 50 && x < 56) printf("MASTER");
    }

}

File: credit.py
import cs50

def main():

    card_number = get_card_number()
    length = len(str(card_number))
    
    if length != 13 and length != 15 and length !=16:
        print("Invalid")  
    
    num_array = []
    for i in range(length):
        num_array.append(card_number % 10)
        card_number = card_number // 10
        
    original_number = num_array.copy()
    temp_number = num_array
    
    for i in range(len(temp_number)):
        if i % 2 != 0:
            temp_number[i] *= 2
   
    total = 0
    if length == 13:
        for i in temp_number:
            total += ((i % 10) + (i // 10 % 10))
        if original_number[12] == 4 and total % 10 == 0:
            print("VISA")
        else:
            print("Invalid_1")
    
    if length == 15:
        for i in temp_number:
            total += ((i % 10) + (i // 10 % 10))
        if (original_number[14] == 3 and total % 10 == 0 and 
           (original_number[13] == 4 or original_number[13] == 7)):
            print("AMEX")
        else:
            print("Invalid_2")
        
    if length == 16:
        for i in temp_number:
            total += ((i % 10) + (i // 10 % 10))
            print(total, end = " ")
        if original_number[15] == 4 and total % 10 == 0:
            print("VISA")    
        elif (original_number[15] == 5 and total % 10 == 0 and 
             (original_number[14] == 1 or original_number[14] == 2 or
             original_number[14] == 3 or original_number[14] == 4 or
             original_number[14] == 5)):
            print("MASTERCARD")
        else:
            print("Invalid_3")
       
    
def get_card_number():
        while True:
            print("Enter card number:")
            card_number = cs50.get_int()
            if card_number > 0:
                break
        return card_number
    
if __name__ == "__main__":
    main()
    
    